raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -261,15 +267,23 @@ private void applyCommittedRecordsToStateMachine() {
     }
 
     private void maybeCommitPendingAppends(LogOffsetMetadata highWatermark, long currentTimeMs) {
-        Iterator<Map.Entry<OffsetAndEpoch, AppendBatchAndTime>> iter = uncommittedAppends.entrySet().iterator();
+        Iterator<Map.Entry<OffsetAndEpoch, UncommittedAppend>> iter = uncommittedAppends.entrySet().iterator();
         while (iter.hasNext()) {
-            Map.Entry<OffsetAndEpoch, AppendBatchAndTime> entry = iter.next();
+            Map.Entry<OffsetAndEpoch, UncommittedAppend> entry = iter.next();
             OffsetAndEpoch offsetAndEpoch = entry.getKey();
             if (offsetAndEpoch.offset < highWatermark.offset) {
-                AppendBatchAndTime appendBatchAndTime = entry.getValue();
-                long elapsedTime = Math.max(0, currentTimeMs - appendBatchAndTime.appendTimeMs);
-                double elapsedTimePerRecord = elapsedTime / (double) appendBatchAndTime.numRecords;
+                UncommittedAppend uncommittedAppend = entry.getValue();
+                long elapsedTime = Math.max(0, currentTimeMs - uncommittedAppend.appendTimeMs);
+                double elapsedTimePerRecord = elapsedTime / (double) uncommittedAppend.numRecords;
                 kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, currentTimeMs);
+
+                // If the commit mode is quorum, then we can complete the future now;
+                // Note that the future may already been expired by the purgatory and in that case,
+                // this is based on the fact that CompletableFuture is thread-safe
+                if (uncommittedAppend.commitMode == AppendCommitMode.QUORUM && !uncommittedAppend.future.isDone()) {
```

Comment:
nit: I'm not sure the `isDone` checks are needed everywhere. Pretty sure `complete` is a no-op if the future has already been completed.

Reply:
Yup, confirmed.